latch_cli/services/cp/utils.py

```diff
@@ -0,0 +1,199 @@
+import os
+import re
+from dataclasses import dataclass
+from enum import Enum
+from pathlib import Path
+from typing import Dict, TypedDict
+from urllib.parse import urlparse
+
+import gql
+
+from latch_cli.config.user import user_config
+from latch_cli.services.cp.exceptions import AuthenticationError, PathResolutionError
+
+
+def get_auth_header() -> Dict[str, str]:
+    sdk_token = user_config.token
+    execution_token = os.environ.get("FLYTE_INTERNAL_EXECUTION_ID")
+
+    if sdk_token is not None and sdk_token != "":
+        headers = {"Authorization": f"Latch-SDK-Token {sdk_token}"}
+    elif execution_token is not None:
+        headers = {"Authorization": f"Latch-Execution-Token {execution_token}"}
+    else:
+        raise AuthenticationError("Unable to find authentication credentials.")
+
+    return headers
+
+
+is_valid_path_expr = re.compile(r"^(latch)?://")
+
+
+def is_remote_path(path: str) -> bool:
+    return is_valid_path_expr.match(path) is not None
+
+
+legacy_expr = re.compile(r"^(account_root|mount)/([^/]+)(?:/+(.*))?$")
+scheme_expr = re.compile(
+    r"^(?:"
+    r"(?P<full>latch://[^/]*(/+.*)?)"
+    r"|(?P<missing_latch>://[^/]*(/+.*)?)"
+    r"|(?P<missing_scheme_with_leading_slash>/+.*)"
+    r"|(?P<missing_scheme_without_leading_slash>[^/]+.*)"
+    r")$"
+)
+domain_expr = re.compile(
+    r"^(((shared\.)?\d+\.account)"  # shared.{acc_id}.account
+    r"|((.+)\.mount)"  # {bucket}.mount
+    r"|(archive)"  # archive
+    r"|((?P<shared_without_selector>shared))"  # shared
+    r"|(\d+\.node))$"  # {node_id}.node
+)
+
+
+# path transform rules:
+#   ://domain/a/b/c => latch://domain/a/b/c
+#   /a/b/c => latch:///a/b/c
+#   a/b/c => latch:///a/b/c
+#
+# domain transform rules:
+#   latch:///a/b/c => latch://xxx.account/a/b/c
+#   latch://shared/a/b/c => latch://shared.xxx.account/a/b/c
+#   latch://any_other_domain/a/b/c => unchanged
+def normalize_path(path: str) -> str:
+    if legacy_expr.match(path):
+        return path  # let nuke-data deal with legacy paths
+
+    match = scheme_expr.match(path)
+    if match is None:
+        raise PathResolutionError(f"{path} is not in a valid format")
+
+    if match["missing_latch"] is not None:
+        path = f"latch{path}"
```

Comment:
It seems unreasonable to accept `://sjdsasdkfjkasddskja`.

Reply:
why?

Reply:
It is an unreasonable thing to type into a command line tool. Keep it if you want but it is confusing that this would be a higher priority than misspelling latch or smtg.